STACpopulator/cli.py

```diff
@@ -1,217 +1,67 @@
-        "--auth-identity",
-        required=False,
-        help="Bearer token, cookie-jar file or proxy/digest/basic username:password for selected authorization handler.",
-    )
-
 
-def apply_request_options(session: Session, namespace: argparse.Namespace) -> None:
-    """
-    Applies the relevant request session options from parsed input arguments.
-    """
-    session.verify = namespace.verify
-    session.cert = namespace.cert
-    if namespace.auth_handler in ["basic", "digest", "proxy"]:
-        usr, pwd = namespace.auth_identity.split(":", 1)
-        if namespace.auth_handler == "basic":
-            session.auth = HTTPBasicAuth(usr, pwd)
-        elif namespace.auth_handler == "digest":
-            session.auth = HTTPDigestAuth(usr, pwd)
-        else:
-            session.auth = HTTPProxyAuth(usr, pwd)
-    elif namespace.auth_handler == "bearer":
-        session.auth = HTTPBearerTokenAuth(namespace.auth_identity)
-    elif namespace.auth_handler == "cookie":
-        session.cookies = HTTPCookieAuth(namespace.auth_identity)
-        session.cookies.load(namespace.auth_identity)
-
-
-def make_main_parser() -> argparse.ArgumentParser:
-    parser = argparse.ArgumentParser(prog="stac-populator", description="STACpopulator operations.")
+def add_parser_args(parser: argparse.ArgumentParser) -> dict[str, Callable]:
     parser.add_argument(
         "--version",
         "-V",
         action="version",
         version=f"%(prog)s {__version__}",
         help="prints the version of the library and exits",
     )
-    commands = parser.add_subparsers(title="command", dest="command", description="STAC populator command to execute.")
-
-    run_cmd_parser = make_run_command_parser(parser.prog)
-    commands.add_parser(
-        "run",
-        prog=f"{parser.prog} {run_cmd_parser.prog}",
-        parents=[run_cmd_parser],
-        formatter_class=run_cmd_parser.formatter_class,
-        usage=run_cmd_parser.usage,
-        add_help=False,
-        help=run_cmd_parser.description,
-        description=run_cmd_parser.description,
+    parser.add_argument("--debug", action="store_const", const=logging.DEBUG, help="set logger level to debug")
+    parser.add_argument(
+        "--log-file", help="file to write log output to. By default logs will be written to the current directory."
     )
+    commands_subparser = parser.add_subparsers(
+        title="command", dest="command", description="STAC populator command to execute.", required=True
+    )
+    run_parser = commands_subparser.add_parser("run", description="Run a STACpopulator implementation")
+    populators_subparser = run_parser.add_subparsers(
+        title="populator", dest="populator", description="Implementation to run."
+    )
+    for implementation_module_name, module in implementation_modules().items():
+        implementation_parser = populators_subparser.add_parser(implementation_module_name)
+        module.add_parser_args(implementation_parser)
```

Comment:
Starting to look into this, and I'm not really fond of the logic depending on a function named "add_parser_args" being present in the module. I don't have a better alternative yet though.

Reply:
The previous approach assumed a `make_parser` instead. The logic just changed the name to represent what the function does, but has similar dependencies.

The only workaround I can think of is to define an abstract `ArgParser` class with a `add_parser_args` method, and each implementation must define their `ArgParser` inheriting from it. The CLI could then look up the modules and filter by `issubclass` of `ArgParser` to find relevant references. That being said, it just defines an explicit structure, but the result would be the same.